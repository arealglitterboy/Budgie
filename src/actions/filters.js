export const setFilterTerm = (term) => ({ type: 'SET_FILTER_TERM', term });

export const sortByNewest = () => ({ type: 'SORT_BY', sortBy: 'byNewest' });

export const sortByOldest = () => ({ type: 'SORT_BY', sortBy: 'byOldest' });

export const sortByAmountDescending = () => ({ type: 'SORT_BY', sortBy: 'byAmountDescending' });

export const sortByAmountAscending = () => ({ type: 'SORT_BY', sortBy: 'byAmountAscending' });

export const setStartDate = (date) => ({ type: 'SET_START_DATE', date });

export const setEndDate = (date) => ({ type: 'SET_END_DATE', date });