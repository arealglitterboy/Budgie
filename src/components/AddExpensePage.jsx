import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses.action';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.addExpense(expense);
        this.props.history.push('/');
    };

    render() {
        return (
            <main>
                <h2>Create Expense</h2>
                <ExpenseForm onSubmit={this.onSubmit} />
            </main>
        );    
    }
}

const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
});

const mapStateToProps = connect(undefined, mapDispatchToProps);

export default mapStateToProps(AddExpensePage);