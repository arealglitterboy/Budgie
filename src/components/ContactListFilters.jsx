'use strict';
import React from 'react';

import { connect } from 'react-redux';
import { setCategories, setFilterTerm, setSortBy } from '../actions/contactsFilters.action';

import SearchInput from './filter/SearchInput';
import CategoriesSelect from './filter/CategoriesSelect';
import SortSelect from './filter/SortSelect';

const option = (value, label) => ({ value, label });

const options = [
    option('byNewest', 'By Newest'),
    option('byOldest', 'By Oldest'),
    option('byNameDescending', 'Name Descending'),
    option('byNameAscending', 'Name Ascending')
];

export const ContactListFilters = (props) => {
    const clearFilters = () => {
        props.setSortBy();
        props.setFilterTerm();
        props.setCategories();
    };

    return (
        <header className="contacts-filters">
            <SearchInput 
                search={props.filters.term}
                setSearch={props.setFilterTerm}
            />
            <SortSelect 
                options={options}
                sortBy={props.filters.sortBy}
                setSortBy={props.setSortBy}
            />
            <CategoriesSelect 
                categories={props.filters.categories}
                setCategories={props.setCategories}
            />
            <button className="contacts-filters__clear" onClick={clearFilters}>clear</button>
        </header>
    );
}


const mapDispatchToProps = (dispatch) => ({
    setFilterTerm: (term) => dispatch(setFilterTerm(term)),
    setSortBy: (sortBy) => dispatch(setSortBy(sortBy)),
    setCategories: (categories) => dispatch(setCategories(categories))
});

const mapStateToProps = connect(({ contactsFilters }) => ({ filters: contactsFilters }), mapDispatchToProps);

export default mapStateToProps(ContactListFilters);