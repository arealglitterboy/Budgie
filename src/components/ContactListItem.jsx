import React from 'react'

const imgOnError = (e) => {
    e.target.src = `data:image/svg+xml;charset=UTF-8,%3csvg xmlns='http://www.w3.org/2000/svg' height='48px' viewBox='0 0 24 24' width='48px' %3e%3cpath d='M0 0h24v24H0V0z' fill='none' /%3e%3cpath d='M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z' /%3e%3c/svg%3e`;
    e.target.classList.add('contact__icon__image--default');
    e.target.onError = () => {};
};

const buttonOnFocus = (e) => {
    console.log();
    e.target.nextSibling.classList.add('contact__categories__list--active')
};

const buttonOnBlur = (e) => {
    e.target.nextSibling.classList.remove('contact__categories__list--active')
}

export default (props) => (
    <article className="contact" id={props.id}>
        <div className="contact__icon">
            <img src={props.icon || ''} alt="" className="contact__icon__image" onError={imgOnError} />
        </div>
        <h5 className="contact__name">{props.name}</h5>
        <p className="contact__date">
            <time dateTime={props.date}>{props.date.toLocaleDateString()}</time>
        </p>
        <section className="contact__categories">
            <button className="contact__categories__button" onFocus={buttonOnFocus} onBlur={buttonOnBlur}>view</button>
            <ul className="contact__categories__list">
                {props.categories.map((val, i) => <li key={i} className="contact__categories__list__item">{val}</li>)}
            </ul>
        </section>
    </article>
)