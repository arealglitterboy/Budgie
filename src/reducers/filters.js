import moment from "moment";

const sort = {
    byNewest: (e1, e2) => (e2.date - e1.date),
    byOldest: (e1, e2) => (e1.date - e2.date),
    byAmountDescending: (e1, e2) => (e2.amount - e1.amount),
    byAmountAscending: (e1, e2) => (e1.amount - e2.amount)
};

const getSort = (by) => ((typeof sort[by] !== 'undefined') ? { sortBy: by, sortFunction: sort[by] } : {});

const defaultState = {
    term: '',
    sortBy: 'byNewest',
    sortFunction: sort['byNewest'],
    startDate: moment().startOf('month'),
    endDate: moment().endOf('month')
};

export default (state = defaultState, action) => {
    switch (action.type) {
        case 'SET_FILTER_TERM':
            return { ...state, term: action.term };
        case 'SORT_BY':
            return { ...state, ...getSort(action.sortBy) };
        case 'SET_START_DATE':
            return { ...state, startDate: action.date };
        case 'SET_END_DATE':
            return { ...state, endDate: action.date };
        case 'SORT_BY_DATE_NEWEST':
            return { ...state, ...getSort('byNewest') }
        case 'SORT_BY_DATE_OLDEST':
            return { ...state, ...getSort('byOldest')};
        case 'SORT_BY_AMOUNT_DESCENDING':
            return { ...state, ...getSort('byAmountDescending') };
        default:
            return state;
    }
};