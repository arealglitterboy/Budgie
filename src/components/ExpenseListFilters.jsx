'use strict';
import React, { useState } from 'react';

import { connect } from 'react-redux';
import { setFilterTerm, setContacts, setSortBy, setEndDate, setStartDate } from '../actions/filters.action';

import ReactDatePicker from 'react-datepicker';

import { isValidDate } from '../utility/validateDates';

import { Input } from './Input';
import InputSelect from './InputSelect';
import InputMultiSelect from './InputMultiSelect';

const option = (value, label) => ({ value, label });

const options = [
    option('byNewest', 'By Newest'),
    option('byOldest', 'By Oldest'),
    option('byAmountDescending', 'Amount Descending'),
    option('byAmountAscending', 'Amount Ascending'),
    option('byTitleDescending', 'Title Descending'),
    option('byTitleAscending', 'Title Ascending')
];

const onDateChange = (onSuccess) => ((input) =>{
    const date = new Date(input);

    if (isValidDate(date)) {
        onSuccess(date);
    }
});

export const ExpenseListFilters = (props) => {
    let [filtersVisible, setFiltersVisible] = useState(false);
    const toggleFiltersVisibility = () => setFiltersVisible(!filtersVisible);
    
    const clearFilters = () => {
        props.setStartDate();
        props.setEndDate();
        props.setSortBy();
        props.setContacts();
    };
    
    const contactToOption = ({id: value, name: label}) => ({ label, value });
    const selectContacts = ({id}) => props.filters.contacts.includes(id);

    const StartDateInput = React.forwardRef(({ value, onClick }, ref) => <Input type="text" label="Start Date" id="set-start-date" ref={ref} value={value} onChange={onDateChange(props.setStartDate)} onClick={onClick} />);
    const EndDateInput = React.forwardRef(({ value, onClick }, ref) => <Input type="text" label="End Date" id="set-end-date" ref={ref} value={value} onChange={onDateChange(props.setEndDate)} onClick={onClick} />);        

    return (
        <header className="filters">

            <section className="filters__top-level">
                <Input className="filters__top-level__search" label="Search" id="search-term" type="search" value={props.filters.term} onChange={props.setFilterTerm} />
                <div>
                    <InputSelect
                        label="Sort"
                        id="sort-expenses"
                        className="filters__top-level__sort"
                        onChange={props.setSortBy}
                        value={options.find(({value}) => value === props.filters.sortBy)}
                        isSearchable={false}
                        isClearable={false}
                        options={options}
                    />
                </div>
                <button className="filters__top-level__button" onClick={toggleFiltersVisibility} data-filters-visible={filtersVisible}>Filters</button>
            </section>

            <section className="filters__options" data-filters-visible={filtersVisible}>
                <div>
                    <ReactDatePicker
                        selectsStart
                        id="start-date"
                        className="filters__options__dates filters__options__dates--start"
                        dateFormat="dd/MM/yyyy"
                        selected={props.filters.startDate}
                        onChange={props.setStartDate}

                        startDate={props.filters.startDate}
                        endDate={props.filters.endDate}

                        maxDate={props.filters.endDate}
                        customInput={<StartDateInput />}
                    />
                </div>
                <div>
                    <ReactDatePicker
                        selectsEnd
                        id="end-date"
                        className="filters__options__dates filters__options__dates--end"
                        dateFormat="dd/MM/yyyy"
                        selected={props.filters.endDate}
                        onChange={props.setEndDate}
                
                        startDate={props.filters.startDate}
                        endDate={props.filters.endDate}
                
                        minDate={props.filters.startDate}
                        customInput={<EndDateInput />}
                    />
                </div>
                <div>
                    <InputMultiSelect 
                        label="Contacts"
                        id="contacts"
                        className="filters__options__contacts"
                        onChange={props.setContacts}
                        value={props.contacts.filter(selectContacts).map(contactToOption)}
                        options={props.contacts.map(contactToOption)}
                    />
                </div>
                <button className="filters__options__clear" onClick={clearFilters}>clear</button>
            </section>
        </header>
    );
}


const mapDispatchToProps = (dispatch) => ({
    setFilterTerm: (term) => dispatch(setFilterTerm(term)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setSortBy: (sortBy) => dispatch(setSortBy(sortBy)),
    setContacts: (contacts) => dispatch(setContacts(contacts))
});

const mapStateToProps = connect(({ filters, contacts }) => ({ filters, contacts }), mapDispatchToProps);

export default mapStateToProps(ExpenseListFilters);