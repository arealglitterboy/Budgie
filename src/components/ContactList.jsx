import React from 'react'
import { connect } from 'react-redux'

import ContactListItem from './ContactListItem';
import selectContacts from '../selectors/contacts.selector';

export const ContactList = ({ contacts=[] }) => (
    <section className="contacts-list">
        {
            (contacts.length === 0) 
            ? <p className="contacts__empty-message">Start by adding</p>
            : contacts.map((contact) => (<ContactListItem {...contact} id={contact.id} name={contact.name} key={contact.id} />))
        }
    </section>
);

const mapStateToProps = connect(({ contacts, contactsFilters }) => ({
    contacts: selectContacts(contacts, contactsFilters)
}));

export default mapStateToProps(ContactList);
