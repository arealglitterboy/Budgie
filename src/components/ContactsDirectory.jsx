'use strict';
import React from 'react'

import ContactListFilters from './ContactListFilters';
import ContactList from './ContactList';

export const ContactsDirectory = (props) => (
    <main className="contacts-directory">
        <h2>Contacts</h2>
        <ContactListFilters />
        <ContactList />
    </main>
);

export default ContactsDirectory
