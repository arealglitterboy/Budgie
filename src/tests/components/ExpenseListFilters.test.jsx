import React from "react";
import { shallow } from "enzyme";

import { ExpenseListFilters } from "../../components/ExpenseListFilters";
import { filters, altFilters } from "../fixtures/filters.fixture";

let setFilterTerm, setSortBy, setEndDate, setStartDate, wrapper;

beforeEach(() => {
    setFilterTerm = jest.fn();
    setSortBy = jest.fn();
    setEndDate = jest.fn();
    setStartDate = jest.fn();
    wrapper = shallow(<ExpenseListFilters
        filters={filters}
        setFilterTerm={setFilterTerm}
        setSortBy={setSortBy}
        setStartDate={setStartDate}
        setEndDate={setEndDate}
    />);
});

test('should render expense list filters correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should render expense list filters with alt data correctly', () => {
    wrapper.setProps({ filters: altFilters });
    expect(wrapper).toMatchSnapshot();
});

test('should handle filter term change', () => {
    const value = 'new search';
    wrapper.find('#search-term').simulate('change', { target: { value } });
    expect(setFilterTerm).toHaveBeenLastCalledWith(value);
});

test('should change sort', () => {
    const select = wrapper.find('#sort-expenses');
    const sorts = ['byOldest', 'byAmountDescending', 'byAmountAscending'];
    
    for (let value of sorts) {
        select.simulate('change', { target: { value } });
        expect(setSortBy).toHaveBeenLastCalledWith(value);
    }
});

test('should handle date change', () => {
    const change = { startDate: new Date("2021-01-04"), endDate: new Date("2021-05-09")};
    wrapper.find('#start-date').prop('onChange')(change.startDate);
    wrapper.find('#end-date').prop('onChange')(change.endDate);
    expect(setStartDate).toHaveBeenLastCalledWith(change.startDate);
    expect(setEndDate).toHaveBeenLastCalledWith(change.endDate);
});