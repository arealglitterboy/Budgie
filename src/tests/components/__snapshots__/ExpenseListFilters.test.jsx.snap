// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list filters correctly 1`] = `
<header
  className="expense-filters"
>
  <section>
    <label
      htmlFor="search-term"
    >
      Search: 
    </label>
    <input
      id="search-term"
      onChange={[Function]}
      type="search"
    />
  </section>
  <section>
    <label
      htmlFor="sort-expense"
    >
      Sort By: 
    </label>
    <select
      id="sort-expenses"
      onChange={[Function]}
    >
      <option
        value="byNewest"
      >
        Newest
      </option>
      <option
        value="byOldest"
      >
        Oldest
      </option>
      <option
        value="byAmountDescending"
      >
        Amount descending
      </option>
      <option
        value="byAmountAscending"
      >
        Amount ascending
      </option>
    </select>
    <section
      className="expense-filters__dates"
    >
      <r
        allowSameDay={false}
        className="expense-filters__dates__input"
        customTimeInput={null}
        dateFormat="dd/MM/yyyy"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        enableTabLoop={true}
        excludeScrollbar={true}
        focusSelectedMonth={false}
        monthsShown={1}
        nextMonthButtonLabel="Next Month"
        nextYearButtonLabel="Next Year"
        onBlur={[Function]}
        onCalendarClose={[Function]}
        onCalendarOpen={[Function]}
        onChange={[Function]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        preventOpenOnFocus={false}
        previousMonthButtonLabel="Previous Month"
        previousYearButtonLabel="Previous Year"
        readOnly={false}
        renderDayContents={[Function]}
        selectsStart={true}
        shouldCloseOnSelect={true}
        showFourColumnMonthYearPicker={false}
        showFullMonthYearPicker={false}
        showMonthYearPicker={false}
        showPopperArrow={true}
        showPreviousMonths={false}
        showQuarterYearPicker={false}
        showTimeInput={false}
        showTimeSelect={false}
        showTwoColumnMonthYearPicker={false}
        showYearPicker={false}
        strictParsing={false}
        timeCaption="Time"
        timeInputLabel="Time"
        timeIntervals={30}
        withPortal={false}
        yearItemNumber={12}
      />
      <r
        allowSameDay={false}
        className="expense-filters__dates__input"
        customTimeInput={null}
        dateFormat="dd/MM/yyyy"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        enableTabLoop={true}
        excludeScrollbar={true}
        focusSelectedMonth={false}
        monthsShown={1}
        nextMonthButtonLabel="Next Month"
        nextYearButtonLabel="Next Year"
        onBlur={[Function]}
        onCalendarClose={[Function]}
        onCalendarOpen={[Function]}
        onChange={[Function]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        preventOpenOnFocus={false}
        previousMonthButtonLabel="Previous Month"
        previousYearButtonLabel="Previous Year"
        readOnly={false}
        renderDayContents={[Function]}
        selectsEnd={true}
        shouldCloseOnSelect={true}
        showFourColumnMonthYearPicker={false}
        showFullMonthYearPicker={false}
        showMonthYearPicker={false}
        showPopperArrow={true}
        showPreviousMonths={false}
        showQuarterYearPicker={false}
        showTimeInput={false}
        showTimeSelect={false}
        showTwoColumnMonthYearPicker={false}
        showYearPicker={false}
        strictParsing={false}
        timeCaption="Time"
        timeInputLabel="Time"
        timeIntervals={30}
        withPortal={false}
        yearItemNumber={12}
      />
    </section>
  </section>
</header>
`;

exports[`should render expense list filters with alt data correctly 1`] = `
<header
  className="expense-filters"
>
  <section>
    <label
      htmlFor="search-term"
    >
      Search: 
    </label>
    <input
      id="search-term"
      onChange={[Function]}
      type="search"
    />
  </section>
  <section>
    <label
      htmlFor="sort-expense"
    >
      Sort By: 
    </label>
    <select
      id="sort-expenses"
      onChange={[Function]}
    >
      <option
        value="byNewest"
      >
        Newest
      </option>
      <option
        value="byOldest"
      >
        Oldest
      </option>
      <option
        value="byAmountDescending"
      >
        Amount descending
      </option>
      <option
        value="byAmountAscending"
      >
        Amount ascending
      </option>
    </select>
    <section
      className="expense-filters__dates"
    >
      <r
        allowSameDay={false}
        className="expense-filters__dates__input"
        customTimeInput={null}
        dateFormat="dd/MM/yyyy"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        enableTabLoop={true}
        excludeScrollbar={true}
        focusSelectedMonth={false}
        monthsShown={1}
        nextMonthButtonLabel="Next Month"
        nextYearButtonLabel="Next Year"
        onBlur={[Function]}
        onCalendarClose={[Function]}
        onCalendarOpen={[Function]}
        onChange={[Function]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        preventOpenOnFocus={false}
        previousMonthButtonLabel="Previous Month"
        previousYearButtonLabel="Previous Year"
        readOnly={false}
        renderDayContents={[Function]}
        selectsStart={true}
        shouldCloseOnSelect={true}
        showFourColumnMonthYearPicker={false}
        showFullMonthYearPicker={false}
        showMonthYearPicker={false}
        showPopperArrow={true}
        showPreviousMonths={false}
        showQuarterYearPicker={false}
        showTimeInput={false}
        showTimeSelect={false}
        showTwoColumnMonthYearPicker={false}
        showYearPicker={false}
        strictParsing={false}
        timeCaption="Time"
        timeInputLabel="Time"
        timeIntervals={30}
        withPortal={false}
        yearItemNumber={12}
      />
      <r
        allowSameDay={false}
        className="expense-filters__dates__input"
        customTimeInput={null}
        dateFormat="dd/MM/yyyy"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        enableTabLoop={true}
        excludeScrollbar={true}
        focusSelectedMonth={false}
        monthsShown={1}
        nextMonthButtonLabel="Next Month"
        nextYearButtonLabel="Next Year"
        onBlur={[Function]}
        onCalendarClose={[Function]}
        onCalendarOpen={[Function]}
        onChange={[Function]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        preventOpenOnFocus={false}
        previousMonthButtonLabel="Previous Month"
        previousYearButtonLabel="Previous Year"
        readOnly={false}
        renderDayContents={[Function]}
        selectsEnd={true}
        shouldCloseOnSelect={true}
        showFourColumnMonthYearPicker={false}
        showFullMonthYearPicker={false}
        showMonthYearPicker={false}
        showPopperArrow={true}
        showPreviousMonths={false}
        showQuarterYearPicker={false}
        showTimeInput={false}
        showTimeSelect={false}
        showTwoColumnMonthYearPicker={false}
        showYearPicker={false}
        strictParsing={false}
        timeCaption="Time"
        timeInputLabel="Time"
        timeIntervals={30}
        withPortal={false}
        yearItemNumber={12}
      />
    </section>
  </section>
</header>
`;
